sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonDataSource.scala

```diff
@@ -92,26 +93,30 @@ object TextInputJsonDataSource extends JsonDataSource {
       sparkSession: SparkSession,
       inputPaths: Seq[FileStatus],
       parsedOptions: JSONOptions): StructType = {
-    val json: Dataset[String] = createBaseDataset(
-      sparkSession, inputPaths, parsedOptions.lineSeparator)
+    val json: Dataset[String] = createBaseDataset(sparkSession, inputPaths, parsedOptions)
+
     inferFromDataset(json, parsedOptions)
   }
 
   def inferFromDataset(json: Dataset[String], parsedOptions: JSONOptions): StructType = {
     val sampled: Dataset[String] = JsonUtils.sample(json, parsedOptions)
-    val rdd: RDD[UTF8String] = sampled.queryExecution.toRdd.map(_.getUTF8String(0))
-    JsonInferSchema.infer(rdd, parsedOptions, CreateJacksonParser.utf8String)
+    val rdd: RDD[InternalRow] = sampled.queryExecution.toRdd
+    val rowParser = parsedOptions.encoding.map { enc =>
+      CreateJacksonParser.internalRow(enc, _: JsonFactory, _: InternalRow, 0)
```

Comment:
Seems the last argument is always 0. Can we omit this?

Reply:
I moved 0 to here to address @cloud-fan review comment why the field by position 0 is taken from InternalRow. It should be clear here why internal rows have only one field at the 0 position. Please, tell me how could omit 0.

Reply:
Can we do something like

```scala
(factory JsonFactory, row: InternalRow) =>
  val bais = new ByteArrayInputStream(row.getBinary(0)))
  CreateJacksonParser.inputStream(enc, factory, bais)
```
?

Looks `internalRow` doesn't actually deduplicate codes.

Reply:
I didn't it originally but rejected the solution because overhead of wrapping the array by `ByteArrayInputStream` per-each row is very high. It increases execution time up to 20% in some cases.

Reply:
sorry I think the array is already wrapped by ByteArrayInputStream per record?